addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusDataThingHandler.java

```diff
@@ -0,0 +1,798 @@
+                ModbusEndpointThingHandler bridgeHandler = (@NonNull ModbusEndpointThingHandler) bridge.getHandler();
+                slaveId = bridgeHandler.getSlaveId();
+                slaveEndpoint = bridgeHandler.asSlaveEndpoint();
+                manager = bridgeHandler.getManagerRef().get();
+                childOfEndpoint = true;
+            } else {
+                @SuppressWarnings("null")
+                @NonNull
+                ModbusPollerThingHandler bridgeHandler = (@NonNull ModbusPollerThingHandler) bridge.getHandler();
+                pollTask = bridgeHandler.getPollTask();
+                if (pollTask == null) {
+                    logger.debug("Poller {} '{}' has no poll task -- configuration is changing?", bridge.getUID(),
+                            bridge.getLabel());
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE,
+                            String.format("Poller %s '%s' has no poll task", bridge.getUID(), bridge.getLabel()));
+                    return;
+                }
+                slaveId = pollTask.getRequest().getUnitID();
+                slaveEndpoint = pollTask.getEndpoint();
+                manager = bridgeHandler.getManagerRef().get();
+                pollStart = pollTask.getRequest().getReference();
+                childOfEndpoint = false;
+            }
+            if (!validateAndParseReadParameters()) {
+                // status already updated to OFFLINE
+                return;
+            }
+            if (!validateAndParseWriteParameters()) {
+                // status already updated to OFFLINE
+                return;
+            }
+            if (!validateMustReadOrWrite()) {
+                // status already updated to OFFLINE
+                return;
+            }
+            updateStatus(ThingStatus.ONLINE);
+        } catch (Throwable e) {
+            logger.error("Exception during initialization", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, String
+                    .format("Exception during initialization: %s (%s)", e.getMessage(), e.getClass().getSimpleName()));
+        } finally {
+            logger.trace("initialize() of thing {} '{}' finished", thing.getUID(), thing.getLabel());
+        }
+    }
+
+    @Override
+    public synchronized void dispose() {
+        readValueType = null;
+        writeValueType = null;
+        readTransformation = null;
+        writeTransformation = null;
+        readIndex = Optional.empty();
+        readSubIndex = Optional.empty();
+        writeStart = null;
+        pollStart = 0;
+        slaveId = 0;
+        slaveEndpoint = null;
+        manager = null;
+        pollTask = null;
+        isWriteEnabled = false;
+        isReadEnabled = false;
+        transformationOnlyInWrite = false;
+        childOfEndpoint = false;
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Thing disposed -- bridge initializing?");
```

Comment:
Indeed this is not necessary because the `ThingManager`

https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java#L775

sets the thing to `UNINITIALIZED` already.